File: Medium/longest_subtstring_without_repeating_characters.php
<?php
class Solution {

    /**
     * @param String $s
     * @return Integer
     */
    function lengthOfLongestSubstring($s) {
        if ($s == "" || strlen($s) > 50000) {
            return 0;
        }

        $longestSubtstring = 1;
        $startIndex = array();
        $startPosition = 0;
        $strSplit = str_split($s);

        for($i = 0; $i < count($strSplit); $i++) {
            if(array_key_exists($strSplit[$i], $startIndex)) {
                if($startIndex[$strSplit[$i]] >= $startPosition) {
                    $startPosition = $startIndex[$strSplit[$i]] + 1;
                }
            }

            $startIndex[$strSplit[$i]] = $i;
            $longestSubtstring = max($longestSubtstring, $i - $startPosition + 1);
        }

        return $longestSubtstring;
    }
}

$solution = new Solution;
print_r($solution->lengthOfLongestSubstring("abcabcbb")); #3
print_r($solution->lengthOfLongestSubstring("bbbbb")); #1
print_r($solution->lengthOfLongestSubstring("pwwkew")); #3
print_r($solution->lengthOfLongestSubstring("")); #0
print_r($solution->lengthOfLongestSubstring("au")); #2
print_r($solution->lengthOfLongestSubstring("aab")); #2

File: Medium/longest_subtstring_without_repeating_characters.py
class Solution(object):
    def lengthOfLongestSubstring(self, s):
        """
        :type s: str
        :rtype: int
        """
        if not s or len(s) > 50000:
            return 0
        
        longestSubtstring = 1
        startIndex = {}
        startPosition = 0
            
        for i, c in enumerate(s):
            if c in startIndex and startIndex[c] >= startPosition:
                startPosition = startIndex[c] + 1
                
            startIndex[c] = i
            longestSubtstring = max(longestSubtstring, i - startPosition + 1)
            
        return longestSubtstring
        
print(Solution().lengthOfLongestSubstring("abcabcbb")) #3
print(Solution().lengthOfLongestSubstring("bbbbb")) #1
print(Solution().lengthOfLongestSubstring("pwwkew")) #3
print(Solution().lengthOfLongestSubstring("")) #0
print(Solution().lengthOfLongestSubstring("au")) #2
print(Solution().lengthOfLongestSubstring("aab")) #2

#NOTE: TIME OUT ERROR on Leetcode 
